src/Nethermind/Nethermind.Evm/Instructions/EvmInstructions.Stack.cs

```diff
@@ -498,19 +505,23 @@ public static EvmExceptionType InstructionPush0<TTracingInst>(VirtualMachine vm,
     /// Executes a PUSH instruction.
     /// Reads immediate data of a fixed length from the code and pushes it onto the stack.
     /// </summary>
+    /// <typeparam name="TGasPolicy">The gas policy implementation.</typeparam>
     /// <typeparam name="TOpCount">The push operation implementation defining the byte count.</typeparam>
+    /// <typeparam name="TTracingInst">The tracing flag.</typeparam>
     /// <param name="vm">The virtual machine instance.</param>
     /// <param name="stack">The execution stack.</param>
-    /// <param name="gasAvailable">The available gas which is reduced by the operation's cost.</param>
+    /// <param name="gasState">The gas state which is reduced by the operation's cost.</param>
     /// <param name="programCounter">Reference to the program counter, which will be advanced.</param>
     /// <returns><see cref="EvmExceptionType.None"/> on success.</returns>
     [SkipLocalsInit]
-    public static EvmExceptionType InstructionPush<TOpCount, TTracingInst>(VirtualMachine vm, ref EvmStack stack, ref long gasAvailable, ref int programCounter)
+    public static EvmExceptionType InstructionPush<TGasPolicy, TOpCount, TTracingInst>(VirtualMachine<TGasPolicy> vm,
+        ref EvmStack stack, ref GasState gasState, ref int programCounter)
+        where TGasPolicy : struct, IGasPolicy<TGasPolicy>
         where TOpCount : IOpCount
         where TTracingInst : struct, IFlag
     {
         // Deduct a very low gas cost for the push operation.
-        gasAvailable -= GasCostOf.VeryLow;
+        TGasPolicy.ConsumeGas(ref gasState, GasCostOf.VeryLow, Instruction.PUSH1 + (byte)(TOpCount.Count - 1));
```

Comment:
`Instruction.PUSH1 + (byte)(TOpCount.Count - 1)` 

should be generalized to some helper method (used in multiple places here) like `TOpCount.GetInstruction(Instruction baseInstruction)`